docs/src/main/sphinx/sql/merge.rst

```diff
@@ -0,0 +1,86 @@
+=====
+MERGE
+=====
+
+Synopsis
+--------
+
+.. code-block:: text
+
+    MERGE INTO target_table [ [ AS ]  target_alias ]
+    USING { source_table | query } [ [ AS ] source_alias ]
+    ON search_condition
+    when_clause [...]
+
+where ``when_clause`` is one of
+
+.. code-block:: text
+
+    WHEN MATCHED [ AND condition ]
+        THEN DELETE
+
+.. code-block:: text
+
+    WHEN MATCHED [ AND condition ]
+        THEN UPDATE SET ( column = expression [, ...] )
+
+.. code-block:: text
+
+    WHEN NOT MATCHED [ AND condition ]
+        THEN INSERT [ column_list ] VALUES (expression, ...)
+
+Description
+-----------
+
+Conditionally update and/or delete rows of a table and/or insert new rows into a table.
+
+``MERGE`` supports an arbitrary number of ``WHEN`` clauses with different ``MATCHED`` conditions,
+executing the ``DELETE``, ``UPDATE`` or ``INSERT`` operation in the first ``WHEN`` clause selected
+by the ``MATCHED`` state and the match condition.
+
+In ``WHEN`` clauses with ``UPDATE`` operations, the column value expressions can depend on any field of the
+target or the source.  In the ``NOT MATCHED`` case, the ``INSERT`` expressions can depend only
+on the source.
+
+Each row in the source must match at most one row in the target table.  If more than one target table
+row is matched by a source row, a ``CONSTRAINT_VIOLATION`` exception is raised.
+
+
+Examples
+--------
+
+Delete all customers mentioned in the source table::
+
+    MERGE INTO target_table t USING source_table s ON (t.customer = s.customer)
+        WHEN MATCHED
+            THEN DELETE
+
+For matching customer rows, increment the purchases, and if there is no match, insert the row
+from the source table::
+
+    MERGE INTO target_table t USING source_table s ON (t.customer = s.customer)
+        WHEN MATCHED
+            THEN UPDATE SET purchases = s.purchases + t.purchases
+        WHEN NOT MATCHED
+            THEN INSERT (customer, purchases, address)
+                  VALUES(s.customer, s.purchases, s.address)
+
+``MERGE`` into the target table from the source table, deleting any matching target row for which
+the source address is Centreville.  For all other matching rows, add the source purchases and
+set the address to the source address, if there is no match in the target table, insert the source
+table row::
+
+    MERGE INTO target_table t USING source_table s ON (t.customer = s.customer)
+        WHEN MATCHED AND s.address = 'Centreville'
+            THEN DELETE
+        WHEN MATCHED
+            THEN UPDATE SET purchases = s.purchases + t.purchases, address = s.address
+        WHEN NOT MATCHED
+            THEN INSERT (customer, purchases, address)
+                  VALUES(s.customer, s.purchases, s.address)
+
+Limitations
+-----------
+
+Some connectors have limited or no support for ``MERGE``.
```

Comment:
I feel like we need a better way to address this than making it an NxM problem.

Reply:
Agreed .. the best would be consistent implementation of such things across all connectors so we just have to call out a few exceptions. Currently I think it would be great to just go in depth on the docs and then use that assembled knowledge to drive consistency in the implementation to clean things up. Currently its all kind of hidden in the code only and thats just really hard for users (and probably even developers when assessing impact of changes..)

Reply:
But all this should be a separate larger effort outside of the scope of this PR for sure..